core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -760,22 +848,10 @@ DynamicFilterContext createContextForQueryRetry(int attemptId)
 
         void addDynamicFilterConsumer(Set<DynamicFilterId> dynamicFilterIds, Consumer<Map<DynamicFilterId, Domain>> consumer)
         {
-            ImmutableMap.Builder<DynamicFilterId, Domain> collectedDomainsBuilder = ImmutableMap.builder();
-            dynamicFilterIds.forEach(dynamicFilterId -> {
-                List<Consumer<Map<DynamicFilterId, Domain>>> consumers;
-                synchronized (dynamicFilterConsumers) {
-                    consumers = dynamicFilterConsumers.get(dynamicFilterId);
-                    if (consumers != null) {
-                        consumers.add(consumer);
-                        return;
-                    }
-                }
-                // filter has already been collected
-                collectedDomainsBuilder.put(dynamicFilterId, dynamicFilterSummaries.get(dynamicFilterId));
-            });
-            Map<DynamicFilterId, Domain> collectedDomains = collectedDomainsBuilder.buildOrThrow();
-            if (!collectedDomains.isEmpty()) {
-                consumer.accept(collectedDomains);
+            for (DynamicFilterId dynamicFilterId : dynamicFilterIds) {
+                DynamicFilterCollectionContext collectionContext = dynamicFilterCollectionContexts.get(dynamicFilterId);
+                verify(collectionContext != null, "collectionContext is missing for %s", dynamicFilterId);
+                addSuccessCallback(collectionContext.getCollectedDomainFuture(), domain -> consumer.accept(ImmutableMap.of(dynamicFilterId, domain)));
```

Comment:
In the previous implementation we had tried to deliver all the dynamic filters that a consumer was interested in within a single batch. This is because the consumer here is going to send dynamic filters to all worker nodes via HttpRemoteTask#sendUpdate. Is it possible to do something like that here as well or do you feel the overheads are small enough that it doesn't matter ?

Reply:
I assume in practice all the filters will be delivered in one shot anyway as it is more likely for the same thread to add more filters before the update is scheduled asynchronously. However I suppose even if it's split into 2 requests it shouldn't be a big deal. We should probably run a benchmark though. But I would be surprised if it had any meaningful impact on performance.